src/net_encryption.cpp

```diff
@@ -0,0 +1,373 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <net_encryption.h>
+
+#include <chainparams.h>
+#include <crypto/hkdf_sha256_32.h>
+#include <net_message.h>
+#include <util.h>
+
+int NetCryptedMessage::Read(const char* pch, unsigned bytes)
+{
+    if (!m_in_data) {
+        // copy data to temporary parsing buffer
+        unsigned int remaining = m_encryption_handler->GetAADLen() - m_hdr_pos;
+        unsigned int copy_bytes = std::min(remaining, bytes);
+
+        memcpy(&vRecv[m_hdr_pos], pch, copy_bytes);
+        m_hdr_pos += copy_bytes;
+
+        // if AAD incomplete, exit
+        if (m_hdr_pos < m_encryption_handler->GetAADLen()) {
+            return copy_bytes;
+        }
+
+        // decrypt the length from the AAD
+        if (!m_encryption_handler->GetLength(vRecv, m_message_size)) {
+            return -1;
+        }
+
+        // check and unset rekey bit
+        // the counterparty can signal a post-this-message rekey by setting the
+        // most significant bit in the (unencrypted) length
+        m_rekey_flag = (m_message_size & (1U << 23));
+        if (m_rekey_flag) {
+            LogPrint(BCLog::NET, "Rekey flag detected %ld\n", m_message_size);
+            m_message_size &= ~(1U << 23);
+        }
+
+        // reject messages larger than MAX_SIZE
+        if (m_message_size > MAX_SIZE) {
+            LogPrint(BCLog::NET, "Max message size exceeded %ld\n", m_message_size);
+            return -1;
+        }
+
+        // switch state to reading message data
+        m_in_data = true;
+
+        return copy_bytes;
+    } else {
+        // copy the message payload plus the MAC tag
+        const unsigned int AAD_LEN = m_encryption_handler->GetAADLen();
+        const unsigned int TAG_LEN = m_encryption_handler->GetTagLen();
+        unsigned int remaining = m_message_size + TAG_LEN - m_data_pos;
+        unsigned int copy_bytes = std::min(remaining, bytes);
+
+        // extend buffer, respect previous copied AAD part
+        if (vRecv.size() < AAD_LEN + m_data_pos + copy_bytes) {
+            // Allocate up to 256 KiB ahead, but never more than the total message size (incl. AAD & TAG).
+            vRecv.resize(std::min(AAD_LEN + m_message_size + TAG_LEN, AAD_LEN + m_data_pos + copy_bytes + 256 * 1024 + TAG_LEN));
+        }
+
+        memcpy(&vRecv[AAD_LEN + m_data_pos], pch, copy_bytes);
+        m_data_pos += copy_bytes;
+
+        if (Complete()) {
```

Comment:
Invert logic here to handle the `!Complete()` case first? That would simplify the code and reduce indentation.